File: src/ts/codeblock-highlight/spec-bound.ts
/**
 * A 1-indexed position, either from the start or end of a line. This mirrors the C# implementation's SpecBound struct, but is a simple
 * interface here since TS has no System.Index/Range to lean on.
 */
export interface SpecBound {
    /**
     * The 1-indexed position from the start or end of a line.
     */
    readonly value: number;

    /**
     * Whether the position is counted from the end of a line (true) or from the start (false).
     */
    readonly isFromEnd: boolean;
}

/**
 * Creates a SpecBound that counts from the start of a line.
 * @param value The 1-indexed position from the start of a line.
 * @returns A SpecBound representing the position.
 */
export function forward(value: number): SpecBound {
    return {value, isFromEnd: false};
}

/**
 * Creates a SpecBound that counts from the end of a line.
 * @param value The 1-indexed position from the end of a line.
 * @returns A SpecBound representing the position.
 */
export function backward(value: number): SpecBound {
    return {value, isFromEnd: true};
}

/**
 * Resolves a SpecBound to a concrete 0-indexed position, given the total length being indexed into
 * (a line count, or a single line's character count).
 * @param bound The SpecBound to resolve.
 * @param role Whether the bound is a start or end position. This affects how the bound is resolved to 0-indexed coordinates.
 * @param length The total length being indexed into (a line count, or a single line's character count).
 * @returns The resolved 0-indexed position.
 */
export function resolveBound(bound: SpecBound, role: 'start' | 'end', length: number): number {
    if (role === 'start') {
        return bound.isFromEnd
            ? length - bound.value // back-start: unchanged relative to length
            : bound.value - 1;     // forward-start: -1 (1-based -> 0-based)
    }

    // role === 'end' (exclusive, for use as an array/slice upper bound)
    return bound.isFromEnd
        ? length - bound.value + 1 // back-end: -1 relative to back-start (inclusive -> exclusive)
        : bound.value;             // forward-end: unchanged (1-based inclusive == 0-based exclusive)
}

/**
 * An inclusive range of 1-indexed bounds.
 */
export interface SpecRange {
    readonly start: SpecBound;
    readonly end: SpecBound;
}

/**
 * Resolves a SpecRange to a concrete [start, end) pair of 0-indexed positions, clamped to [0, length].
 * @param range The SpecRange to resolve.
 * @param length The total length being indexed into (a line count, or a single line's character count).
 * @returns An object containing the resolved start and end positions.
 */
export function resolveRange(range: SpecRange, length: number): { start: number; end: number } {
    const start = clamp(resolveBound(range.start, 'start', length), 0, length);
    const end = clamp(resolveBound(range.end, 'end', length), 0, length);
    return {start, end: Math.max(start, end)};
}

function clamp(value: number, min: number, max: number): number {
    return Math.min(Math.max(value, min), max);
}

File: BoothDotDev.CodeBlockTrivia/Highlighting/SpecBound.cs
namespace BoothDotDev.CodeBlockTrivia.Highlighting;

/// <summary>
///     Represents a single 1-indexed bound, either from the start or from the end, as parsed from a highlight trivia spec
///     (e.g. the <c>2</c> in <c>2..8</c>, or the <c>2</c> in <c>^2</c>).
/// </summary>
/// <param name="Value">The 1-indexed value of the bound.</param>
/// <param name="IsFromEnd">
///     <see langword="true" /> if this bound counts from the end (i.e. was written with a <c>^</c> prefix); otherwise,
///     <see langword="false" />.
/// </param>
public readonly record struct SpecBound(int Value, bool IsFromEnd)
{
    /// <summary>
    ///     Creates a <see cref="SpecBound" /> counting from the start (forward), e.g. spec <c>2</c>.
    /// </summary>
    /// <param name="value">The 1-indexed value.</param>
    /// <returns>A new <see cref="SpecBound" /> instance.</returns>
    public static SpecBound Forward(int value)
    {
        return new SpecBound(value, IsFromEnd: false);
    }

    /// <summary>
    ///     Creates a <see cref="SpecBound" /> counting from the end (backward), e.g. spec <c>^2</c>.
    /// </summary>
    /// <param name="value">The 1-indexed value, counted from the end.</param>
    /// <returns>A new <see cref="SpecBound" /> instance.</returns>
    public static SpecBound Backward(int value)
    {
        return new SpecBound(value, IsFromEnd: true);
    }

    /// <summary>
    ///     Converts this bound to a <see cref="System.Index" />, treating it as the <b>start</b> of a range.
    /// </summary>
    /// <returns>The equivalent <see cref="System.Index" />.</returns>
    public Index ToStartIndex()
    {
        return IsFromEnd
            ? new Index(Value, fromEnd: true) // spec back-index is already aligned with Index's fromEnd semantics
            : new Index(Value - 1); // spec is 1-based, Index is 0-based
    }

    /// <summary>
    ///     Converts this bound to a <see cref="System.Index" />, treating it as the <b>end</b> of a range.
    /// </summary>
    /// <returns>The equivalent <see cref="System.Index" />.</returns>
    public Index ToEndIndex()
    {
        return IsFromEnd
            ? new Index(Value - 1, fromEnd: true) // spec end is inclusive, Range.End is exclusive
            : new Index(Value); // 1-based inclusive-end coincides with 0-based exclusive-end
    }
}
